### ISSUE TITLE
[no-invalid-void-type] Reports Promise<string | void> as error

### ISSUE THREAD

**Issue by avalanche1** [typescript-eslint/typescript-eslint#3014]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-invalid-void-type": "error"
  }
}
```

```TS
browser.runtime.onMessage.addListener(
  async (message: string): Promise<string | void> => {
      if (message !== 'DO_IT') return;
  
     return 'done';
  };
)
```

**Expected Result**
No errors.
Because I'm returning early w\o any actual return value; because the listener is only supposed to return a value _only_ if he gets a proper command; otherwise he must ignore it. 
With `void` return type I am intent on telling the compiler that the listener might not return any result at all and just quit.

**Actual Result**
ESLint: void is only valid as a return type or generic type variable

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^4.14.1` |
| `@typescript-eslint/parser`        | `^4.14.1` |
| `TypeScript`                       | `^4.1.3` |
| `ESLint`                           | `^7.19.0` |

**Comment by bradzacher** [comment]
The entire point of the rule is to prevent you from doing things like this.

The rule enforces that `void` is only ever used on its own.

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-invalid-void-type.md#rationale

In this instance you do not have a `void` function, you have a function that implicitly returns `undefined` in one branch.

Eg use `string | void`.
